1. Importing Libraries

In [ ]:
import numpy as np
import pandas as pd

2. Loading Data

In [ ]:
c=pd.read_excel('Dataset.xlsx',sheet_name='CustomerDemographic')

3. Checking records

In [ ]:
c.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases                  DOB  \
0                                   93  1953-10-12 00:00:00   
1                                   81  1980-12-16 00:00:00   
2                                   61  1954-01-20 00:00:00   
3                                   33  1961-10-03 00:00:00   
4                                   56  1977-05-13 00:00:00   

                job_title job_industry_category     wealth_segment  \
0     Executive Secretary                Health      Mass Customer   
1  Administrative Officer    Financial Services      Mass Customer   
2      Recruiting Manager              Property      Mass Customer   
3                     NaN                    IT      Mass Customer   
4           Senior Editor                   NaN  Affluent Customer   

  deceased_indicator                                            default  \
0                  N                                                 "'   
1                  N                       <script>alert('hi')</script>   
2                  N                                2018-02-01 00:00:00   
3                  N  () { _; } >_[$($())] { touch /tmp/blns.shellsh...   
4                  N                                                NIL   

  owns_car  tenure  
0      Yes    11.0  
1      Yes    16.0  
2      Yes    15.0  
3       No     7.0  
4      Yes     8.0

4. Check information of columns and their datatypes

In [ ]:
c.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   customer_id                          4000 non-null   int64  
 1   first_name                           4000 non-null   object 
 2   last_name                            3875 non-null   object 
 3   gender                               4000 non-null   object 
 4   past_3_years_bike_related_purchases  4000 non-null   int64  
 5   DOB                                  3913 non-null   object 
 6   job_title                            3494 non-null   object 
 7   job_industry_category                3344 non-null   object 
 8   wealth_segment                       4000 non-null   object 
 9   deceased_indicator                   4000 non-null   object 
 10  default                              3698 non-null   object 
 11  owns_car                      

Data information:RangeIndex: 4000 entries, 0 to 3999
Data columns=13
By checking head and information we find out that column 10 ie. default is irrelevant which shopuld be deleted from dataset

1. Dropping Irrelevent Column

In [ ]:
c.drop(labels={'default'}, axis=1 , inplace=True)

2. Handling Missing Value

In [ ]:
#returns the number of missing values in the dataset
c.isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                  87
dtype: int64

In [ ]:
# Find percentage of missing values in each column of a #pandas dataframe
c.isnull().mean()*100

customer_id                             0.000
first_name                              0.000
last_name                               3.125
gender                                  0.000
past_3_years_bike_related_purchases     0.000
DOB                                     2.175
job_title                              12.650
job_industry_category                  16.400
wealth_segment                          0.000
deceased_indicator                      0.000
owns_car                                0.000
tenure                                  2.175
dtype: float64

**Note:** Column last_name, DOB, job_title, job_indsutry_category, tenure has missing values

2.1 Last Name


In [ ]:
# Checking where last name is missing.
c[c['last_name'].isnull()]

customer_id first_name last_name  gender  \
3               4     Talbot       NaN    Male   
66             67     Vernon       NaN    Male   
105           106       Glyn       NaN    Male   
138           139        Gar       NaN    Male   
196           197       Avis       NaN  Female   
...           ...        ...       ...     ...   
3884         3885      Asher       NaN    Male   
3915         3916     Myrtia       NaN  Female   
3926         3927     Conway       NaN    Male   
3961         3962     Benoit       NaN    Male   
3998         3999  Patrizius       NaN    Male   

      past_3_years_bike_related_purchases                  DOB  \
3                                      33  1961-10-03 00:00:00   
66                                     67  1960-06-14 00:00:00   
105                                    54  1966-07-03 00:00:00   
138                                     1  1964-07-28 00:00:00   
196                                    32  1977-01-27 00:00:00   
...                                   ...                  ...   
3884                                   55  1978-06-17 00:00:00   
3915                                   31  1958-10-17 00:00:00   
3926                                   29  1978-01-07 00:00:00   
3961                                   17  1977-10-06 00:00:00   
3998                                   11  1973-10-24 00:00:00   

                       job_title job_industry_category     wealth_segment  \
3                            NaN                    IT      Mass Customer   
66              Web Developer II                Retail      Mass Customer   
105   Software Test Engineer III                Health     High Net Worth   
138                     Operator    Telecommunications  Affluent Customer   
196                          NaN                   NaN     High Net Worth   
...                          ...                   ...                ...   
3884                     Actuary    Financial Services      Mass Customer   
3915                         NaN                Retail  Affluent Customer   
3926         Electrical Engineer         Manufacturing      Mass Customer   
3961             Project Manager           Argiculture     High Net Worth   
3998                         NaN         Manufacturing  Affluent Customer   

     deceased_indicator owns_car  tenure  
3                     N       No     7.0  
66                    N       No    18.0  
105                   N      Yes    18.0  
138                   N       No     4.0  
196                   N       No     5.0  
...                 ...      ...     ...  
3884                  N      Yes     8.0  
3915                  N      Yes    17.0  
3926                  N      Yes     7.0  
3961                  N      Yes    14.0  
3998                  N      Yes    10.0  

[125 rows x 12 columns]

Note: We have 125 rows of missing last_name value

In [ ]:
c['last_name'].isnull().sum()

125

In [ ]:
#Checking the values of customer id and first name corresponding to missing last_name
c[c['last_name'].isnull()][[ 'customer_id','first_name']].isnull().sum()

customer_id    0
first_name     0
dtype: int64

**Note:** We observe that the values where last_name is missing has both customer_id and first_name which makes it recognizable even in the absence of missing value of last_name.
**Solution:** filling missing value of last_name(125 values) with Null

In [ ]:
c['last_name'].fillna('None',axis=0, inplace=True)

In [ ]:
#confirming updated Missing value
c['last_name'].isnull().sum()

0

2.2 DOB

In [ ]:
# Checking where DOB is missing.
c[c['DOB'].isnull()]

customer_id first_name   last_name gender  \
143           144       Jory  Barrabeale      U   
167           168     Reggie   Broggetti      U   
266           267      Edgar     Buckler      U   
289           290    Giorgio      Kevane      U   
450           451     Marlow   Flowerdew      U   
...           ...        ...         ...    ...   
3778         3779      Ulick      Daspar      U   
3882         3883      Nissa      Conrad      U   
3930         3931      Kylie       Epine      U   
3934         3935     Teodor   Alfonsini      U   
3997         3998     Sarene     Woolley      U   

      past_3_years_bike_related_purchases  DOB                   job_title  \
143                                    71  NaN          Environmental Tech   
167                                     8  NaN             General Manager   
266                                    53  NaN                         NaN   
289                                    42  NaN      Senior Sales Associate   
450                                    37  NaN  Quality Control Specialist   
...                                   ...  ...                         ...   
3778                                   68  NaN                         NaN   
3882                                   35  NaN             Legal Assistant   
3930                                   19  NaN                         NaN   
3934                                   72  NaN                         NaN   
3997                                   60  NaN           Assistant Manager   

     job_industry_category     wealth_segment deceased_indicator owns_car  \
143                     IT      Mass Customer                  N       No   
167                     IT  Affluent Customer                  N      Yes   
266                     IT     High Net Worth                  N       No   
289                     IT      Mass Customer                  N       No   
450                     IT     High Net Worth                  N       No   
...                    ...                ...                ...      ...   
3778                    IT  Affluent Customer                  N       No   
3882                    IT      Mass Customer                  N       No   
3930                    IT     High Net Worth                  N      Yes   
3934                    IT     High Net Worth                  N      Yes   
3997                    IT     High Net Worth                  N       No   

      tenure  
143      NaN  
167      NaN  
266      NaN  
289      NaN  
450      NaN  
...      ...  
3778     NaN  
3882     NaN  
3930     NaN  
3934     NaN  
3997     NaN  

[87 rows x 12 columns]

In [ ]:
c['DOB'].isnull().sum()

87

to check what percentage of DOB is null.
case 1: if percentage is low we drop it
case 2: find alternative

In [ ]:
c['DOB'].isnull().mean()*100

2.175

In [ ]:
dob_drop= c[c['DOB'].isnull()].index
dob_drop

Index([ 143,  167,  266,  289,  450,  452,  453,  479,  512,  525,  547,  581,
        598,  679,  684,  798,  838,  882,  891,  949,  974,  982,  995, 1037,
       1043, 1081, 1173, 1209, 1243, 1350, 1476, 1508, 1582, 1627, 1682, 1739,
       1772, 1779, 1805, 1917, 1937, 1989, 1999, 2020, 2068, 2164, 2204, 2251,
       2294, 2334, 2340, 2413, 2425, 2468, 2539, 2641, 2646, 2695, 2696, 2853,
       2919, 2962, 2998, 3011, 3085, 3150, 3221, 3222, 3254, 3287, 3297, 3311,
       3321, 3342, 3364, 3472, 3509, 3512, 3564, 3653, 3717, 3726, 3778, 3882,
       3930, 3934, 3997],
      dtype='int64')

In [ ]:
c.drop(index=dob_drop, inplace=True, axis=0)

In [ ]:
#checking current DOB value
c['DOB'].isnull().sum()

0

2.2.2 There can be case where age can be invalid,but we do not have age column so we create a function age to find out current age and remove any invalid age if found

In [ ]:
c['DOB'].info()

<class 'pandas.core.series.Series'>
Index: 3913 entries, 0 to 3999
Series name: DOB
Non-Null Count  Dtype 
--------------  ----- 
3913 non-null   object
dtypes: object(1)
memory usage: 61.1+ KB


In [ ]:
from datetime import datetime

def calc_age(born):
    today = datetime.today()
    return today.year - born.year - ((today.month, today.day) < (born.month, born.day))

#There was an error in DOB column where the column DOB contails string and not datetime objects.
#So we converted DOB to datetime format
#5 DOB 3913 non-null   datetime64[ns]
c['DOB'] = pd.to_datetime(c['DOB'])

#create a new column named age and feeding it with calc_age
c['Age'] = c['DOB'].apply(calc_age)

#later you can check DOB datatype changes from **object** to **datetime64[ns]**

In [ ]:
#to get description of age
c['Age'].describe()

count    3913.000000
mean       46.241503
std        12.799428
min        22.000000
25%        37.000000
50%        46.000000
75%        56.000000
max       180.000000
Name: Age, dtype: float64

We found out that max age is 180 which is invalid so we need to remove it

In [ ]:
age_drop =c[c['Age']>100].index
c.drop(index=age_drop, inplace=True , axis=0)

**During Subtask find out:
a) age group chart
b) age group of customers that are newly associated**

2.3 job_title

In [ ]:
# Checking where job_title is missing.
c[c['job_title'].isnull()]

customer_id first_name  last_name  gender  \
3               4     Talbot       None    Male   
5               6       Curr  Duckhouse    Male   
6               7       Fina     Merali  Female   
10             11      Uriah     Bisatt    Male   
21             22    Deeanne   Durtnell  Female   
...           ...        ...        ...     ...   
3967         3968  Alexandra      Kroch  Female   
3971         3972  Maribelle   Schaffel  Female   
3978         3979      Kleon       Adam    Male   
3986         3987     Beckie    Wakeham  Female   
3998         3999  Patrizius       None    Male   

      past_3_years_bike_related_purchases        DOB job_title  \
3                                      33 1961-10-03       NaN   
5                                      35 1966-09-16       NaN   
6                                       6 1976-02-23       NaN   
10                                     99 1954-04-30       NaN   
21                                     79 1962-12-10       NaN   
...                                   ...        ...       ...   
3967                                   99 1977-12-22       NaN   
3971                                    6 1979-03-28       NaN   
3978                                   67 1974-07-13       NaN   
3986                                   18 1964-05-29       NaN   
3998                                   11 1973-10-24       NaN   

     job_industry_category     wealth_segment deceased_indicator owns_car  \
3                       IT      Mass Customer                  N       No   
5                   Retail     High Net Worth                  N      Yes   
6       Financial Services  Affluent Customer                  N      Yes   
10                Property      Mass Customer                  N       No   
21                      IT      Mass Customer                  N       No   
...                    ...                ...                ...      ...   
3967              Property     High Net Worth                  N       No   
3971                Retail      Mass Customer                  N       No   
3978    Financial Services      Mass Customer                  N      Yes   
3986           Argiculture      Mass Customer                  N       No   
3998         Manufacturing  Affluent Customer                  N      Yes   

      tenure  Age  
3        7.0   62  
5       13.0   57  
6       11.0   48  
10       9.0   69  
21      11.0   61  
...      ...  ...  
3967    22.0   46  
3971     8.0   45  
3978    18.0   49  
3986     7.0   59  
3998    10.0   50  

[497 rows x 13 columns]

In [ ]:
c['job_title'].isnull().sum()

497

In [ ]:
c['job_title'].isnull().mean()*100

12.704498977505112

In [ ]:
#Since job_title missing is 12.70 percentage we replace null value with the word "missing"
c['job_title'].fillna('missing', inplace=True, axis=0)

In [ ]:
#rechecking job_title value
c['job_title'].isnull().sum()

0

2.3 job_industry_category

In [ ]:
c[c['job_industry_category'].isnull()]

customer_id      first_name last_name  gender  \
4               5  Sheila-kathryn    Calton  Female   
7               8             Rod     Inder    Male   
15             16          Harlin      Parr    Male   
16             17           Heath   Faraday    Male   
17             18          Marjie   Neasham  Female   
...           ...             ...       ...     ...   
3965         3966          Astrix   Sigward  Female   
3973         3974           Misha   Ranklin  Female   
3975         3976          Gretel  Chrystal  Female   
3982         3983          Jarred     Lyste    Male   
3999         4000           Kippy   Oldland    Male   

      past_3_years_bike_related_purchases        DOB             job_title  \
4                                      56 1977-05-13         Senior Editor   
7                                      31 1962-03-30       Media Manager I   
15                                     38 1977-02-27      Media Manager IV   
16                                     57 1962-03-19       Sales Associate   
17                                     79 1967-07-06             Professor   
...                                   ...        ...                   ...   
3965                                   53 1968-09-15           Geologist I   
3973                                   82 1961-02-11      Technical Writer   
3975                                    0 1957-11-20      Internal Auditor   
3982                                   19 1965-04-21      Graphic Designer   
3999                                   76 1991-11-05  Software Engineer IV   

     job_industry_category     wealth_segment deceased_indicator owns_car  \
4                      NaN  Affluent Customer                  N      Yes   
7                      NaN      Mass Customer                  N       No   
15                     NaN      Mass Customer                  N      Yes   
16                     NaN  Affluent Customer                  N      Yes   
17                     NaN  Affluent Customer                  N       No   
...                    ...                ...                ...      ...   
3965                   NaN      Mass Customer                  N      Yes   
3973                   NaN  Affluent Customer                  N      Yes   
3975                   NaN  Affluent Customer                  N      Yes   
3982                   NaN      Mass Customer                  N      Yes   
3999                   NaN  Affluent Customer                  N       No   

      tenure  Age  
4        8.0   46  
7        7.0   62  
15      18.0   47  
16      15.0   62  
17      11.0   56  
...      ...  ...  
3965    11.0   55  
3973     9.0   63  
3975    13.0   66  
3982     9.0   58  
3999    11.0   32  

[656 rows x 13 columns]

In [ ]:
c['job_industry_category'].isnull().mean()*100

16.768916155419223

In [ ]:
#Since job_industry_category missing is 16.76 percentage we replace null value with the word "missing"
c['job_industry_category'].fillna('missing', inplace=True, axis=0)

In [ ]:
c['job_industry_category'].isnull().sum()

0

#3. Data Inconsistency

Data can be inconsistent in non-numeric section so we need toc heck what are the non-numeric column


In [ ]:
non_numeric = c.select_dtypes(exclude=[np.number])
non_numeric_col = non_numeric.columns.values
non_numeric_col

array(['first_name', 'last_name', 'gender', 'DOB', 'job_title',
       'job_industry_category', 'wealth_segment', 'deceased_indicator',
       'owns_car'], dtype=object)

So we check each field : gender, job-title_job_industry_category_wealth_segment_deceased_indicator,owns_car

3.1 Gender

In [ ]:
c['gender'].value_counts()

gender
Female    2037
Male      1872
F            1
Femal        1
M            1
Name: count, dtype: int64

In [ ]:
def updated_gender(gender):
    if gender=='M':
        return 'Male'
    elif gender=='F':
        return 'Female'
    elif gender=='Femal':
        return 'Female'
    else :
        return gender
c['gender'] = c['gender'].apply(updated_gender)

In [ ]:
c['gender'].value_counts()

gender
Female    2039
Male      1873
Name: count, dtype: int64

3.2 job_industry_category

In [ ]:
c['job_industry_category'].value_counts()

job_industry_category
Manufacturing         796
Financial Services    767
missing               656
Health                596
Retail                358
Property              267
IT                    151
Entertainment         136
Argiculture           113
Telecommunications     72
Name: count, dtype: int64

3.3 job_title

In [ ]:
c['job_title'].value_counts()

job_title
missing                                 497
Business Systems Development Analyst     43
Tax Accountant                           43
Social Worker                            42
Recruiting Manager                       41
                                       ... 
Database Administrator II                 4
Health Coach III                          3
Health Coach I                            3
Research Assistant III                    3
Developer I                               1
Name: count, Length: 196, dtype: int64

3.4 Wealth Segment

In [ ]:
c['wealth_segment'].value_counts()

wealth_segment
Mass Customer        1954
High Net Worth        996
Affluent Customer     962
Name: count, dtype: int64

3.5 Deaseced_indicator

In [ ]:
c['deceased_indicator'].value_counts()

deceased_indicator
N    3910
Y       2
Name: count, dtype: int64

3.6 Owns_car

In [ ]:
c['owns_car'].value_counts()

owns_car
Yes    1974
No     1938
Name: count, dtype: int64

#4. Duplicate

since customer_id is primary then we ned to check if customer _id is uniue or not. if duplicate is found we remove it


In [ ]:
# Check for duplicate rows based on specific columns
df=pd.DataFrame(c)
duplicate_rows_subset = df.duplicated(subset=['customer_id'])

# Check if there are any duplicates
if duplicate_rows_subset.any():
    print("There are duplicates based on 'customer_id'.")
else:
    print("There are no duplicates based on 'customer_id'.")


There are no duplicates based on 'customer_id'.


In [ ]:
c.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3912 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          3912 non-null   int64         
 1   first_name                           3912 non-null   object        
 2   last_name                            3912 non-null   object        
 3   gender                               3912 non-null   object        
 4   past_3_years_bike_related_purchases  3912 non-null   int64         
 5   DOB                                  3912 non-null   datetime64[ns]
 6   job_title                            3912 non-null   object        
 7   job_industry_category                3912 non-null   object        
 8   wealth_segment                       3912 non-null   object        
 9   deceased_indicator                   3912 non-null   object        
 10  owns_car         

#5. Exporting

c.to_csv('CustomerDemographic_cleaned.csv', index=False)

In [ ]:
c.to_csv('CustomerDemographic_cleaned.csv', index=False)